drivers/mtd_flashpage/mtd_flashpage.c

```diff
@@ -26,87 +28,128 @@
 #include "architecture.h"
 #include "cpu.h"
 #include "cpu_conf.h"
+#include "macros/utils.h"
 #include "mtd_flashpage.h"
 #include "periph/flashpage.h"
 
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
 #define MTD_FLASHPAGE_END_ADDR     ((uint32_t) CPU_FLASH_BASE + (FLASHPAGE_NUMOF * FLASHPAGE_SIZE))
 
 static int _init(mtd_dev_t *dev)
 {
-    (void)dev;
+    mtd_flashpage_t *super = container_of(dev, mtd_flashpage_t, base);
+    (void)super;
     assert(dev->pages_per_sector * dev->page_size == FLASHPAGE_SIZE);
+    assert(!(super->offset % dev->pages_per_sector));
+    assert(super->offset + dev->pages_per_sector * dev->page_size * dev->sector_count
+           <= MTD_FLASHPAGE_END_ADDR);
+    assert(super->offset + dev->pages_per_sector * dev->page_size * dev->sector_count
+           >= dev->pages_per_sector * dev->page_size * dev->sector_count);
     return 0;
 }
 
-static int _read(mtd_dev_t *dev, void *buf, uint32_t addr, uint32_t size)
+static int _read_page(mtd_dev_t *dev, void *buf, uint32_t page,
+                      uint32_t offset, uint32_t size)
 {
-    assert(addr < MTD_FLASHPAGE_END_ADDR);
+    mtd_flashpage_t *super = container_of(dev, mtd_flashpage_t, base);
+
+    assert(page + super->offset >= page);
+
```

Comment:
I think you accidently removed `page += super->offset;`

Reply:
Oops, good catch